include/deal.II/lac/petsc_snes.h

```diff
@@ -0,0 +1,477 @@
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Upon returning, the @p x vector contains the solution.
+     */
+    unsigned int
+    solve(VectorType &x, const VectorType &b);
+
+    /**
+     * Solve the affine nonlinear system of equations $F(x) = b$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Here we also set the matrix to handle the Jacobian.
+     */
+    unsigned int
+    solve(VectorType &x, const VectorType &b, PMatrixType &P);
+
+    /**
+     * Solve the affine nonlinear system of equations $F(x) = b$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Here we also set the matrices to handle Jacobians.
+     */
+    unsigned int
+    solve(VectorType &x, const VectorType &b, AMatrixType &A, PMatrixType &P);
+
+    /**
+     * Callback for the computation of the nonlinear residual $F(x)$.
+     */
+    std::function<int(const VectorType &x, VectorType &res)> residual;
+
+    /**
+     * Callback for the computation of the Jacobian
+     * $\dfrac{\partial F}{\partial x}$
+     */
+    std::function<int(const VectorType &x, AMatrixType &A, PMatrixType &P)>
+      jacobian;
+
+    /**
+     * Callback for monitoring the solution process.
+     *
+     * This function is called by NonlinearSolver at the beginning
+     * of each time step.
+     */
+    std::function<
+      int(const VectorType &x, const unsigned int step_number, real_type fval)>
+      monitor;
+
+    /**
+     * Setup Jacobian callback without matrices.
+     *
+     * This callback gives full control to users to set up the tangent
+     * operator $\dfrac{\partial F}{\partial x}$
+     *
+     * Solvers must be provided via NonlinearSolver::solve_for_jacobian_system.
+     */
+    std::function<int(const VectorType &x)> setup_jacobian;
+
+    /**
+     * Solution of the Jacobian system set up with
+     * NonlinearSolver::setup_jacobian
+     *
+     * This is used as a preconditioner inside the Krylov process.
+     */
+    std::function<int(const VectorType &src, VectorType &dst)>
+      solve_for_jacobian_system;
+
+    /**
+     * Callback for the computation of the energy function.
+     *
+     * This is usually not needed, since by default SNES assumes that the
+     * objective function to be minimized is $\frac{1}{2} || F(x) ||^2 $.
+     *
+     * However, if the nonlinear equations are derived from energy arguments, it
+     * may be useful to use this callback to perform linesearch or to test for
+     * the reduction in a trust region step.
+     */
+    std::function<int(const VectorType &x, real_type &obj)> energy;
+
+  protected:
```

Comment:
Private?

Reply:
I usually prefer to keep implementation details `protected` so that advanced users can inherit from these classes and access the implementation details.

Reply:
`protected` means that you have to maintain a specific interface not just via the `public` members, but also the `protected` members. For all practical purposes, if you use `protected`, *everything* becomes `public`. You can't even change the names of internal member variables any more. So let's make things `private` instead.